### PR TITLE
added -del command to delete output from all host

### PR DESCRIPTION
**Pull request by psnbaba**
Use case: when user creates some files which are dumped outside the input_dir of a simulations. e.g. the pre-processing of a geometry creates a file accessible to all by dumping the final output in the parent directory. In this case it is not deleted automatically.  Use can pass this command with the host name to delete the entire outputs. This does not affect the localhost files.

### CODE REVIEW

File: automan/automation.py [hunk 1]
```diff
@@ -1070,5 +1073,10 @@ def _setup_argparse(self):
             dest='update_remote', default=False,
             help='Update remote worker machines.'
         )
+        parser.add_argument(
+            '-del', '--delete-outputs', action='store_true',
```
Comment (prabhuramachandran) [inline automan/automation.py]: I do not like the `-del`, instead use a single character for the single-dash case.  Also is `--delete-remote` what this means?  Please think of a more suitable name that describes what is being done.

File: automan/cluster_manager.py [hunk 2]
```diff
@@ -368,6 +377,14 @@ def update(self, rebuild=True):
                 if rebuild:
                     self._rebuild(host, home)
 
+    def delete(self, sim_dir):
+        for worker in self.workers:
+            host = worker.get('host')
+            home = worker.get('home')
+            nfs = worker.get('nfs', False)
+            if host != 'localhost' and not nfs:
```
Comment (prabhuramachandran) [inline automan/cluster_manager.py]: What about locally?

Comment (psnbaba) [reply automan/cluster_manager.py]: user can delete the folders in local machine manually, but for remote he has to go to each location to delete the output folders incase he need to do so.

File: automan/automation.py (hunk 1 above)
Comment (psnbaba) [reply automan/automation.py]: I can't think of a suitable single character argument, So can we just keep `--delete-outputs`.

**Comment by prabhuramachandran** [comment]
Please resolve the conflicts and also add some tests for this new functionality.

File: automan/cluster_manager.py [hunk 3]
```diff
@@ -326,6 +326,15 @@ def _update_sources(self, host, home):
         )
         self._run_command(cmd)
 
+    def _delete_outputs(self, host, home, sim_dir):
+        path = os.path.join(home, self.root, self.project_name,
+                            sim_dir)
+        real_host = '' if self.testing else '{host}'.format(host=host)
```
Comment (prabhuramachandran) [inline automan/cluster_manager.py:332]: This will wipe out local files in the testing case?  Is that what you intend?

File: automan/cluster_manager.py [hunk 4]
```diff
@@ -368,6 +377,26 @@ def update(self, rebuild=True):
                 if rebuild:
                     self._rebuild(host, home)
 
+    def delete(self, sim_dir, remotes):
+        if remotes[-1] == 'all':
+            hosts = [h.get('host') for h in self.workers]
+            remotes = hosts.copy()
```
Comment (prabhuramachandran) [inline automan/cluster_manager.py:383]: Why is this needed?  The previous line has a copy of the hosts anyway?

File: automan/cluster_manager.py [hunk 5]
```diff
@@ -368,6 +377,26 @@ def update(self, rebuild=True):
...
+        else:
+            for remote in remotes:
+                hosts = [h.get('host') for h in self.workers]
```
Comment (prabhuramachandran) [inline automan/cluster_manager.py:386]: Why is this in the for loop?  the workers don't change in this scope.

File: automan/tests/test_cluster_manager.py [hunk 6]
```diff
@@ -187,6 +187,34 @@ def test_remote_bootstrap_and_sync(self):
...
+        # Given
+        cm = ClusterManager()
+        cm.add_worker('host', home='/home/foo', nfs=False)
+        cm.add_worker('host1', home='/home/foo1', nfs=False)
+
+        # When
+        cm.delete('outputs', ['host', 'host1'])
+
+        # Then
+        mock_delete_outputs.assert_called_with('host1', '/home/foo1', 'outputs')
```
Comment (prabhuramachandran) [inline automan/tests/test_cluster_manager.py:202]: Why is this only called once?  Shouldn't it be called with host and host1?

File: automan/tests/test_cluster_manager.py [hunk 7]
```diff
@@ -187,6 +187,34 @@ def test_remote_bootstrap_and_sync(self):
...
+        # Given
+        cm = ClusterManager()
+        cm.add_worker('host', home='/home/foo', nfs=False)
+        cm.add_worker('host1', home='/home/foo1', nfs=False)
+
+        # When
+        cm.delete('outputs', ['all'])
+
+        # Then
+        mock_delete_outputs.assert_called_with('host1', '/home/foo1', 'outputs')
```
Comment (prabhuramachandran) [inline automan/tests/test_cluster_manager.py:216]: Same here?  In the case of 'all' it should wipe out all hosts.

Also you should test the case where nfs is True.

File: automan/cluster_manager.py (hunk 3 above)
Comment (psnbaba) [reply automan/cluster_manager.py:332]: Yes, It skips localhost and the worker having nfs=True.

File: automan/cluster_manager.py (hunk 4 above)
Comment (psnbaba) [reply automan/cluster_manager.py:383]: Since remote list is required later in the function. When 'all' is passed it copies worker list from the original host list.

File: automan/cluster_manager.py (hunk 5 above)
Comment (psnbaba) [reply automan/cluster_manager.py:386]: fixed.

File: automan/tests/test_cluster_manager.py (hunk 6 above)
Comment (psnbaba) [reply automan/tests/test_cluster_manager.py:202]: fixed.

File: automan/tests/test_cluster_manager.py (hunk 7 above)
Comment (psnbaba) [reply automan/tests/test_cluster_manager.py:216]: fixed and added test when nfs==True

File: automan/automation.py [hunk 8]
```diff
@@ -1174,6 +1177,12 @@ def _setup_argparse(self):
             dest='update_remote', default=False,
             help='Update remote worker machines.'
         )
+        parser.add_argument(
+            '--rm-remote-output', nargs='*', action="store",
+            dest="rm_remote_output", type=str, default='',
+            help="remove output folder from the mentioned machines"
+            "use 'all' to remove from all host"
```
Comment (prabhuramachandran) [inline automan/automation.py:1184]: Will this also wipe out any local output?  When you say "all" is localhost included?  Please mention this in the help documentation here.

File: automan/cluster_manager.py [hunk 9]
```diff
@@ -368,6 +377,25 @@ def update(self, rebuild=True):
...
 
+    def delete(self, sim_dir, remotes):
+        hosts = [h.get('host') for h in self.workers]
+        if remotes[-1] == 'all':
+            remotes = hosts.copy()
+        else:
+            for remote in remotes:
+                if remote not in hosts:
+                    print('%s remote is not a worker' % (remote))
+                    sys.exit()
```
Comment (prabhuramachandran) [inline automan/cluster_manager.py:388]: Use an explicit exit status here, 1 for error usually.

File: automan/tests/test_cluster_manager.py [hunk 10]
```diff
@@ -187,6 +187,50 @@ def test_remote_bootstrap_and_sync(self):
...
+        # When
+        cm.delete('outputs', ['host', 'host1'])
+
+        # Then
+        mock_delete_outputs.assert_any_call('host1', '/home/foo1', 'outputs')
+        mock_delete_outputs.assert_any_call('host', '/home/foo', 'outputs')
+
+    @mock.patch.object(ClusterManager, '_bootstrap')
+    @mock.patch.object(ClusterManager, '_delete_outputs')
+    def test_do_not_delete_outputs_from_hosts_when_nfs_true(self, mock_delete_outputs, mock_bootstrap):
```
Comment (prabhuramachandran) [inline automan/tests/test_cluster_manager.py:207]: Make sure the code fits in 78 cols.

File: automan/cluster_manager.py (hunk 9 above)
Comment (psnbaba) [reply automan/cluster_manager.py:388]: done

File: automan/tests/test_cluster_manager.py (hunk 10 above)
Comment (psnbaba) [reply automan/tests/test_cluster_manager.py:207]: done

File: automan/automation.py (hunk 8 above)
Comment (psnbaba) [reply automan/automation.py:1184]: done
